mycroft/util/audio_utils.py

```diff
@@ -0,0 +1,199 @@
+# Copyright 2020 Mycroft AI Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""Contains simple tools for performing audio related tasks such as playback
+of audio, recording and listing devices.
+"""
+from copy import deepcopy
+import os
+import pyaudio
+import re
+import subprocess
+
+import mycroft.configuration
+
+from .string_utils import get_http
+from .log import LOG
+
+
+def play_audio_file(uri: str, environment=None):
+    """Play an audio file.
+
+    This wraps the other play_* functions, choosing the correct one based on
+    the file extension. The function will return directly and play the file
+    in the background.
+
+    Arguments:
+        uri:    uri to play
+        environment (dict): optional environment for the subprocess call
+
+    Returns: subprocess.Popen object. None if the format is not supported or
+             an error occurs playing the file.
+    """
+    extension_to_function = {
+        '.wav': play_wav,
+        '.mp3': play_mp3,
+        '.ogg': play_ogg
+    }
+    _, extension = os.path.splitext(uri)
+    play_function = extension_to_function.get(extension.lower())
+    if play_function:
+        return play_function(uri, environment)
+    else:
+        LOG.error("Could not find a function capable of playing {uri}."
+                  " Supported formats are {keys}."
+                  .format(uri=uri, keys=list(extension_to_function.keys())))
+        return None
+
+
+_ENVIRONMENT = deepcopy(os.environ)
+_ENVIRONMENT['PULSE_PROP'] = 'media.role=music'
+
+
+def _get_pulse_environment(config):
+    """Return environment for pulse audio depeding on ducking config."""
+    tts_config = config.get('tts', {})
+    if tts_config and tts_config.get('pulse_duck'):
+        return _ENVIRONMENT
+    else:
+        return os.environ
```

Comment:
We don't want to change the global environ but the commands should be launched with a modified environment. We use a static copy to avoid copying every single time.

Reply:
Why don't we want to change the global environment?  Is there some breakage that occurs if we do?  If this code remains, I would at least like to see some comments around why this approach was taken.

Also, can we move the _ENVIRONMENT attribute into the _get_pulse_environment function?

Reply:
I can definitely add a comment.

Breaking is a quite hard term but the audio behavior will be changed for the entire Mycroft processes and all their future subprocess. We just want the commands we're launching through these methods to have the specified PULSE_PROP making pulse duck them when things like TTS is executed.

I do not want to have it inside the function since the copy action is quite heavy and I do not want it done on every call. Since it will be the same each time it's used it's kept at module level.

Reply:
I added the comment explaining the use of the modified environment a bit more.